data/pfd.1
.TH PFD "1" "February 2019" "pfd" "User Commands"
.SH NAME
pfd \- convert font description between xft and pango formats
.SH SYNOPSIS
.B pfd
.RI [ OPTIONS ] 
.RI FONT
.SH DESCRIPTION
\fBpfd\fR is a small command-line utility to transform font names from pango to xft specification and vice versa.
.SH OPTIONS
.TP
\fB\-x\fR, \fB\-\-xft\fR
Print font name in xft format. This is default.
.TP
\fB\-p\fR, \fB\-\-pango\fR
Print font name in pango format.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help.
.TP
\fB\-v\fR, \fB\-\-version\fR
Print current version.


NEWS
Version 0.42.0
    - add pfd utility for fontnames transformation in scripts
    - add 'menu' action for notification icon
    - add --icon-size option to icon dialog
    - add --simple-tips option to list dialog
    - fixes in setting window size
    - improve size scaling in separate output of font dialog

Version 0.41.0
    - add supoort for webkit2
    - use https as a default protocol in html dialog
    - html dialog in browser mode treats extra argument as url to load
    - add --expand option to a notebook dialog
    - improve gtk3 dialog
    - many bug fixes (details in git changelog)

Version 0.40.3
    - shows overwrite confirmation dialog for save file dialog only on even button codes

Version 0.40.2
    - fix --image-path option
    - fix typos in man page

Version 0.40.0
    - add --exit-on-drop argument to dnd dialog
    - use --response option to set default exit code
    - add --iec-format option for size values
    - add --escape-ok option
    - add FormFeed handling in icon dialog
    - use %s in command for dnd dialog
    - fix parse numeric arguments in entry dialog
    - add fix of progress log from BrandonH <bhaberfeld@gmail.com>
    - add --icon-size option to notification dialog

Version 0.39.0
    - add --active-tab to notebook widget
    - place value position of scale form field to the left
    - add --uri-handler option to html dialog
    - add --theme option to text-info dialog (works with GtkSourceView)
    - add --user-agent and --user-style options to html dialog
    - remove --parent-win option
    - add --add-on-top and --tail options to list dialog

Version 0.38.2
    - fix enter behavior

Version 0.38.1
    - fix tooltip option in dnd dialog
    - fix russian and ukrainian translations

Version 0.38.0
    - add ctrl+enter behavior for all dialogs
    - add --inc-button option to scale dialog
    - fix --no-escape handling
    - add --show-week option to calendar dialog
    - improve behavior of --filename option of file dialog

Version 0.37.0
    - new layout management (fix gtk3 warning about missing parent)
    - fix single-click selection in list dialog
    - add --grid-lines option to list dialog
    - add --posx and --posy options for setting window position
    - fix quoting and escaping arguments for handlers in list and form dialogs
    - add wrapping ability for columns in list dialog
    - wrapped, ellipsized and editable columns may be specified as list
    - add 'perm' type of bar in multiprogress dialog

Version 0.36.3
    - fix segfault when parse desktop files with missed or empty Comment field
      in icons dialog
    - code and output cleanup

Version 0.36.2
    - honor precision for floats in numeric fields of form and entry dialogs
    - fix settings title and icon for browser mode in html dialog

Version 0.36.1
    - fix quoting buttons commands in form dialog

Version 0.36.0
    - add --no-selection option to list dialog
    - improve --center option
    - add spell checking support for text field in form and text-info dialog
    - add gtksourceview support to text-info dialog
    - fix setting initial value for numeric text entry in entry dialog
    - do not close print dialog when preview button is clicked
    - add --monitor option to icon dialog
    - change default font for text-info and print text files to Monospace
    - quote field values for button callback in form dialog

Version 0.35.0
    - add BAR column type to list dialog
    - add SZ column type to list dialog (displays human readable sizes)
    - add --show-cursor option to text dialog
    - fix handling leading spaces in button callback of form dialog
    - fix --show-hidden behavior
    - switch float output from %g to %f
    - fix setting custom bg color in text-info dialog for gtk3

Version 0.34.2
    - fix clearing fields in form when --cycle-read is using

Version 0.34.1
    - fix expander
    - update man page

Version 0.34.0
    - add --select-action option to list dialog
    - add --separate-output option to font dialog
    - honor --quoted-output option in font dialog
    - add --show-hidden option to file selection dialogs
    - add --watch-bar to multi-progress dialog
    - improve handling of autoclosing event in multi-progress dialog
    - improve output of color values in form dialog
    - fix build with gtk+-3.0 and webkit-3.0

Version 0.33.1
    - fix timeout handling

Version 0.33.0
    - add --picture dialog, a simple image viewer
    - add --focus-field option to form dialog for set initialy fosuced field
    - add --complete option for extended completions in entry fields
    - add ability to fill form data through stdin
    - fix handling empty data in form dialog
    - improvements in browser mode of html dialog
    - add --float-precision option
    - improve output of floating point numbers

Version 0.32.0
    - enable rules-hint for icons list in icon browser
    - add --splash options (thanks to Konstantin Baierer)
    - add --no-focus option
    - fix timeout progressbar layout for gtk3 builds
    - add --num-output option for output index of active element in combo-boxes

Version 0.31.3
    - fix setting default value for completion entry in form dialog

Version 0.31.2
    - fix segfault in list dialog when separators enabled but --sep-value is unset

Version 0.31.1
    - fix segfault in html dialog

Version 0.31.0
    - add separators to list dialog
    - use YAD_OPTIONS variable for specify default yad arguments
    - add --no-rules-hint option to list dialog
    - add parameters for controlling palettes in color selection
    - remove some entries from settings file
    - set default open command and date format through the settings file
    - add additional filter settings for file selection dialogs
    - apply filter and preview settings to all of file selection dialogs
    - fix settings custom icon theme

Version 0.30.1
    - fix parsing paned orientation

Version 0.30.0
    - add paned dialog

src/pfd.c
/*
 * This file is part of YAD.
 *
 * YAD is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * YAD is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with YAD. If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2019, Victor Ananjevsky <ananasik@gmail.com>
 */

#include <config.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <pango/pango.h>

typedef enum {
  PANGO_SPEC,
  XFT_SPEC
} FontType;

static FontType font_type = XFT_SPEC;

static gboolean
set_type (const gchar *name, const gchar *val, gpointer d, GError **err)
{
  gint i = 1;

  if (name[i] == '-')
    i = 2;

  if (name[i] == 'p')
    font_type = PANGO_SPEC;
  else if (name[i] == 'x')
    font_type = XFT_SPEC;
  else
    return FALSE;

  return TRUE;
}

static gchar *
parse_font (gchar *fn)
{
  gchar **fnt;
  gint i = 1;
  GString *pfd = g_string_new (NULL);

  fnt = g_strsplit (fn, ":",  -1);

  if (g_ascii_strcasecmp (fnt[0], "xft") != 0)
    return fnt[0];

  while (fnt[i])
    {
      if (g_ascii_strncasecmp (fnt[i], "style=", 6) == 0)
        g_string_append_printf (pfd, "%s ", fnt[i] + 6);
      else if (g_ascii_strncasecmp (fnt[i], "size=", 5) == 0)
        g_string_append_printf (pfd, "%s ", fnt[i] + 5);
      else if (g_ascii_strncasecmp (fnt[i], "pixelsize=", 10) == 0)
        g_string_append_printf (pfd, "%spx ", fnt[i] + 10);
      else
        g_string_append_printf (pfd, "%s ", fnt[i]);
      i++;
    }

  return pfd->str;
}

static gchar **fonts = NULL;
static gboolean ver = FALSE;
static GOptionEntry ents[] = {
  { "xft", 'x', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_type, N_("Print font name in xft format"), NULL },
  { "pango", 'p', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_type, N_("Print font name in pango format"), NULL },
  { "version", 'v', 0, G_OPTION_ARG_NONE, &ver, N_("Print version"), NULL },
  { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_STRING_ARRAY, &fonts, 0, N_("FONT") },
  { NULL }
};

int
main (int argc, char *argv[])
{
  GOptionContext *ctx;
  GError *err = NULL;
  PangoFontDescription *fnd = NULL;

  setlocale (LC_ALL, "");

#ifdef ENABLE_NLS
  bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);
#endif

  ctx = g_option_context_new (_("- convert font description"));
  g_option_context_add_main_entries (ctx, ents, GETTEXT_PACKAGE);
  if (!g_option_context_parse (ctx, &argc, &argv, &err))
    {
      g_printerr (_("option parsing failed: %s\n"), err->message);
      return 1;
    }

  if (ver)
    {
      g_print ("%s\n", VERSION);
      return 0;
    }

  if (!fonts || !fonts[0])
    {
      g_printerr (_("no font specified\n"));
      return 1;
    }

  /* parse font */
  fnd = pango_font_description_from_string (parse_font (fonts[0]));
  if (!fnd)
    {
      g_printerr (_("cannot get font description for for string \"%s\"\n"), fonts[0]);
      return 1;
    }

  /* print font */
  switch (font_type)
    {
    case PANGO_SPEC:
      g_print ("%s\n", pango_font_description_to_string (fnd));
      break;
    case XFT_SPEC:
      {
        PangoFontMask mask = pango_font_description_get_set_fields (fnd);

        if (mask & PANGO_FONT_MASK_FAMILY)
          g_print ("xft:%s", pango_font_description_get_family (fnd));
        else
          g_print ("xft:Sans"); /* simple default */

        if (mask & PANGO_FONT_MASK_WEIGHT)
          {
            switch (pango_font_description_get_weight (fnd))
              {
              case PANGO_WEIGHT_THIN:
                g_print (":Thin");
                break;
              case PANGO_WEIGHT_ULTRALIGHT:
                g_print (":Ultralight");
                break;
              case PANGO_WEIGHT_LIGHT:
                g_print (":Light");
                break;
              case PANGO_WEIGHT_SEMILIGHT:
                g_print (":Semilight");
                break;
              case PANGO_WEIGHT_BOOK:
                g_print (":Book");
                break;
              case PANGO_WEIGHT_NORMAL:
                break;
              case PANGO_WEIGHT_MEDIUM:
                g_print (":Medium");
                break;
              case PANGO_WEIGHT_SEMIBOLD:
                g_print (":Semibold");
                break;
              case PANGO_WEIGHT_BOLD:
                g_print (":Bold");
                break;
              case PANGO_WEIGHT_ULTRABOLD:
                g_print (":Ultrabold");
                break;
              case PANGO_WEIGHT_HEAVY:
                g_print (":Heavy");
                break;
              case PANGO_WEIGHT_ULTRAHEAVY:
                g_print (":Ultraheavy");
                break;
              }
          }

        if (mask & PANGO_FONT_MASK_STYLE)
          {
            switch (pango_font_description_get_style (fnd))
              {
              case PANGO_STYLE_NORMAL:
                break;
              case PANGO_STYLE_OBLIQUE:
                g_print (":Oblique");
                break;
              case PANGO_STYLE_ITALIC:
                g_print (":Italic");
                break;
              }
          }

        if (mask & PANGO_FONT_MASK_SIZE)
          {
            if (pango_font_description_get_size_is_absolute (fnd))
              g_print (":pixelsize=%d", pango_font_description_get_size (fnd) / PANGO_SCALE);
            else
              g_print (":size=%d", pango_font_description_get_size (fnd) / PANGO_SCALE);
          }

        g_print ("\n");

        break;
      }
    }

  return 0;
}
